File: golang/internal/datasource/rabbitmq.go
package datasource

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type RabbitMQResponse struct {
	Items []struct {
		Name     string `json:"name"`
		Messages int    `json:"messages"`
	} `json:"items"`
}

func GetRabbitMQQueueLen(host, port, user, pass, pattern string) (map[string]int, error) {
	safePattern := url.QueryEscape(pattern)
	apiURL := fmt.Sprintf("http://%s:%s/api/queues?page=1&page_size=500&name=%s&use_regex=true&pagination=true", host, port, safePattern)

	client := &http.Client{
		Timeout: 10 * time.Second,
	}

	req, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		return nil, err
	}
	req.SetBasicAuth(user, pass)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var rabbitResp RabbitMQResponse
	if err := json.NewDecoder(resp.Body).Decode(&rabbitResp); err != nil {
		return nil, err
	}

	result := make(map[string]int)
	for _, item := range rabbitResp.Items {
		result[item.Name] = item.Messages
	}

	return result, nil
}

File: rust/src/data_source/rabbitmq.rs
use reqwest::Client;
use serde::Deserialize;
use std::collections::HashMap;
use url::form_urlencoded;

#[derive(Deserialize)]
struct RabbitQueue {
    name: String,
    messages: u64,
}

#[derive(Deserialize)]
struct RabbitResponse {
    items: Vec<RabbitQueue>,
}

pub async fn get_queue_lengths(
    host: &str,
    port: &str,
    user: &str,
    pass: &str,
    pattern: &str,
) -> Result<HashMap<String, u64>, Box<dyn std::error::Error>> {
    let client = Client::new();
    let safe_pattern: String = form_urlencoded::byte_serialize(pattern.as_bytes()).collect();
    
    let url = format!(
        "http://{}:{}/api/queues?page=1&page_size=500&name={}&use_regex=true&pagination=true",
        host, port, safe_pattern
    );

    let res = client
        .get(&url)
        .basic_auth(user, Some(pass))
        .send()
        .await?
        .json::<RabbitResponse>()
        .await?;

    let mut ret = HashMap::new();
    for queue in res.items {
        ret.insert(queue.name, queue.messages);
    }

    Ok(ret)
}
